### PR TITLE
feat: support inverted index

### PR DESCRIPTION
**Pull request by longjiquan**
issue: https://github.com/milvus-io/milvus/issues/27704

Add inverted index for some data types in Milvus. This index type can save a lot of memory compared to loading all data into RAM and speed up the term query and range query.

Supported: `INT8`, `INT16`, `INT32`, `INT64`, `FLOAT`, `DOUBLE`, `BOOL` and `VARCHAR`.

Not supported: `ARRAY` and `JSON`.

Note:
- The inverted index for `VARCHAR` is not designed to serve full-text search now. We will treat every row as a whole keyword instead of tokenizing it into multiple terms.
- The inverted index don't support retrieval well, so if you create inverted index for field, those operations which depend on the raw data will fallback to use chunk storage, which will bring some performance loss. For example, comparisons between two columns and retrieval of output fields.

The inverted index is very easy to be used.

Taking below collection as an example:

```python
fields = [
		FieldSchema(name="pk", dtype=DataType.VARCHAR, is_primary=True, auto_id=False, max_length=100),
		FieldSchema(name="int8", dtype=DataType.INT8),
		FieldSchema(name="int16", dtype=DataType.INT16),
		FieldSchema(name="int32", dtype=DataType.INT32),
		FieldSchema(name="int64", dtype=DataType.INT64),
		FieldSchema(name="float", dtype=DataType.FLOAT),
		FieldSchema(name="double", dtype=DataType.DOUBLE),
		FieldSchema(name="bool", dtype=DataType.BOOL),
		FieldSchema(name="varchar", dtype=DataType.VARCHAR, max_length=1000),
		FieldSchema(name="random", dtype=DataType.DOUBLE),
		FieldSchema(name="embeddings", dtype=DataType.FLOAT_VECTOR, dim=dim),
]
schema = CollectionSchema(fields)
collection = Collection("demo", schema)
```

Then we can simply create inverted index for field via:

```python
index_type = "INVERTED"
collection.create_index("int8", {"index_type": index_type})
collection.create_index("int16", {"index_type": index_type})
collection.create_index("int32", {"index_type": index_type})
collection.create_index("int64", {"index_type": index_type})
collection.create_index("float", {"index_type": index_type})
collection.create_index("double", {"index_type": index_type})
collection.create_index("bool", {"index_type": index_type})
collection.create_index("varchar", {"index_type": index_type})
```

Then, term query and range query on the field can be speed up automatically by the inverted index:

```python
result = collection.query(expr='int64 in [1, 2, 3]', output_fields=["pk"])
result = collection.query(expr='int64 < 5', output_fields=["pk"])
result = collection.query(expr='int64 > 2997', output_fields=["pk"])
result = collection.query(expr='1 < int64 < 5', output_fields=["pk"])
```

### CODE REVIEW

File: internal/core/thirdparty/tantivy/tantivy-binding/src/array.rs [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+}
+
+#[no_mangle]
+pub extern "C" fn free_rust_array(array: RustArray) {
+    defer_on_unwind!({
+        ::std::process::abort();
+    });
+    let RustArray { array, len } = array;
+    unsafe {
+        drop::<Box<[u32]>>(Box::from_raw(slice::from_raw_parts_mut(
```
Comment (chyezh) [inline internal/core/thirdparty/tantivy/tantivy-binding/src/array.rs:32]: use Vec::from_raw_parts should be ok.

File: internal/core/thirdparty/tantivy/tantivy-binding/src/array.rs [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+}
+
+impl RustArray {
+    pub fn from_vec(vec: Vec<u32>) -> RustArray {
+        defer_on_unwind!({
+            ::std::process::abort();
+        });
+        let v: Box<[u32]> = vec.into_boxed_slice();
+        RustArray {
+            len: v.len().try_into().expect("Integer Overflow"),
```
Comment (chyezh) [inline internal/core/thirdparty/tantivy/tantivy-binding/src/array.rs:19]: Vec has capacity properties. use len as here, may cause bug in some memory allocator.

File: internal/core/thirdparty/tantivy/tantivy-binding/src/index.rs [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+use tantivy::directory::MmapDirectory;
+use tantivy::query::{Query, RangeQuery, TermQuery, RegexQuery};
+use tantivy::schema::{Field, IndexRecordOption, Schema, TextFieldIndexing, TextOptions, INDEXED};
+use tantivy::{doc, tokenizer, DocAddress, Index, IndexReader, IndexWriter, ReloadPolicy, Term};
+
+use crate::data_type::TantivyDataType;
+use crate::util::make_bounds;
+
+pub struct IndexWrapper {
+    pub field_name: Option<String>,
```
Comment (chyezh) [inline internal/core/thirdparty/tantivy/tantivy-binding/src/index.rs:17]: all those field should never be none by current implementation.
Remove the Option<*> is a better implementation for IndexWrapper.

File: internal/core/thirdparty/tantivy/tantivy-binding/src/index.rs [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+    pub index_writer: Option<IndexWriter>,
+    pub reader: Option<IndexReader>,
+}
+
+impl IndexWrapper {
+    fn create_reader(&mut self) {
+        let reader = self
+            .index
+            .as_mut()
+            .unwrap()
```
Comment (chyezh) [inline internal/core/thirdparty/tantivy/tantivy-binding/src/index.rs:31]: Too much unwrap. 
Remove Option in IndexWrapper definition and change create_reader as an associated function but not a method of IndexWrapper.

create_reader() -> IndexReader

File: internal/core/thirdparty/tantivy/tantivy-binding/src/index.rs [hunk 5]
```diff
@@ -0,0 +1,329 @@
...
+            sum += meta.max_doc();
+        }
+        sum
+    }
+
+    fn search(&mut self, q: &dyn Query) -> Vec<u32> {
+        let searcher = self.reader.as_mut().unwrap().searcher();
+        let cnt = self.count();
+        let hits = searcher
+            .search(q, &TopDocs::with_limit(cnt as usize))
```
Comment (chyezh) [inline internal/core/thirdparty/tantivy/tantivy-binding/src/index.rs:186]: Does search always returns Ok() at runtime?
no Err() handling but unwrap will cause the program crash at runtime.

File: internal/core/thirdparty/tantivy/tantivy-binding/src/index.rs [hunk 6]
```diff
@@ -0,0 +1,329 @@
...
+            index: Some(index),
+            index_writer: None,
+            reader: None,
+        };
+        ret.create_reader();
+        ret.reload();
+        ret
+    }
+
+    pub fn add_i8(&mut self, data: i8) {
```
Comment (chyezh) [inline internal/core/thirdparty/tantivy/tantivy-binding/src/index.rs:110]: it seems that all `add_*` method is immutable method, use `&self` is enough.

File: internal/core/thirdparty/tantivy/tantivy-binding/src/mod.rs
[the remark this answers is not in the document]
Comment (longjiquan) [reply internal/core/thirdparty/tantivy/tantivy-binding/src/mod.rs:1]: removed

File: internal/core/thirdparty/tantivy/tantivy-binding/src/array.rs [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+#[repr(C)]
+pub struct RustArray {
+    array: *mut u32,
+    len: size_t,
+}
+
+impl RustArray {
+    pub fn from_vec(vec: Vec<u32>) -> RustArray {
+        defer_on_unwind!({
+            ::std::process::abort();
```
[the remark this answers is not in the document]
Comment (longjiquan) [reply internal/core/thirdparty/tantivy/tantivy-binding/src/array.rs]: removed

File: internal/core/thirdparty/tantivy/tantivy-binding/src/array.rs (hunk 2 above)
Comment (longjiquan) [reply internal/core/thirdparty/tantivy/tantivy-binding/src/array.rs]: fixed.

File: internal/core/thirdparty/tantivy/tantivy-binding/src/index_writer.rs [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+            }
+            TantivyDataType::Keyword => {
+                let text_field_indexing = TextFieldIndexing::default()
+                    .set_tokenizer("raw_tokenizer")
+                    .set_index_option(IndexRecordOption::Basic);
+                let text_options = TextOptions::default().set_indexing_options(text_field_indexing);
+                field = schema_builder.add_text_field(&field_name, text_options);
+                use_raw_tokenizer = true;
+            }
+            _ => {
```
Comment (chyezh) [inline internal/core/thirdparty/tantivy/tantivy-binding/src/index_writer.rs:52]: remove U64 Text in enum definition to avoid use _ to match all.

Comment (longjiquan) [reply internal/core/thirdparty/tantivy/tantivy-binding/src/index_writer.rs:52]: OK.

**Comment by longjiquan** [comment]
rerun ut

File: internal/core/src/index/IndexFactory.cpp [hunk 9]
```diff
@@ -25,14 +25,22 @@
...
 namespace milvus::index {
 
 template <typename T>
 ScalarIndexPtr<T>
 IndexFactory::CreateScalarIndex(
     const IndexType& index_type,
-    const storage::FileManagerContext& file_manager_context) {
+    const storage::FileManagerContext& file_manager_context,
+    DataType d_type) {
+    if (index_type == "inverted_index") {
```
Comment (czs007) [inline internal/core/src/index/IndexFactory.cpp:38]: rename to INVERTED and define a const ?

File: internal/proxy/task_index.go [hunk 10]
```diff
@@ -160,6 +160,8 @@ func (cit *createIndexTask) parseIndexParams() error {
 			if exist && !validateArithmeticIndexType(specifyIndexType) {
 				return merr.WrapErrParameterInvalid(DefaultArithmeticIndexType, specifyIndexType, "index type not match")
 			}
+		} else if typeutil.IsBoolType(cit.fieldSchema.DataType) {
+			// TODO: should we check something?
```
Comment (czs007) [inline internal/proxy/task_index.go:164]: how about :
if !exist {
				indexParamsMap[common.IndexTypeKey] = INVERTED_INDEX_TYPE
			}
			if exist && specifyIndexType!= INVERTED_INDEX_TYPE) {
				return merr.WrapErrParameterInvalid(INVERTED_INDEX_TYPE, specifyIndexType, "index type not match")
			}

File: internal/core/src/index/IndexFactory.cpp (hunk 9 above)
Comment (longjiquan) [reply internal/core/src/index/IndexFactory.cpp:38]: OK.

**Pull request merged**